USoC'23: Update session 04 - Binary Compatibility (#262)
unikraft/docs: content/en/community/hackathons/sessions/bincompat/content/Work-Items_00.-Setup.md

> To setup, build and run Linux ELFs with [`app-elfloader`](https://github.com/unikraft/app-elfloader), we recommend you use [the `run-app-elfloader` repository](https://github.com/unikraft/run-app-elfloader).

Reviewer: ```suggestion
To set up, build and run Linux ELFs with [`app-elfloader`](https://github.com/unikraft/app-elfloader), we recommend you use [the `run-app-elfloader` repository](https://github.com/unikraft/run-app-elfloader).
```

> Clone [the `run-app-elfloader` repository](https://github.com/unikraft/run-app-elfloader) on your machine to get started.

Reviewer: Remove this line, to make this compliant to the other instructions.

$ git clone https://github.com/unikraft/static-pie-apps
$ git clone https://github.com/unikraft/dynamic-apps

$ cd run-app-elfloader/

$ ./run_app.sh helloworld

SeaBIOS (version rel-1.16.2-0-gea1b7a073390-prebuilt.qemu.org)
Booting from ROM..TEST nofollow
Powered by
o.   .o       _ _               __ _
Oo   Oo  ___ (_) | __ __  __ _ ' _) :_
oO   oO ' _ `| | |/ /  _)' _` | |_|  _)
oOo oOO| | | | |   (| | | (_) |  _) :_
 OoOoO ._, ._:_:_,\_._,  .__,_:_, \___)
                  Atlas 0.13.1~d20aa7cb
[...]
brk(NULL) = va:0x400200000
brk(va:0x400221000) = va:0x400221000
Hello, World!
write(fd:1, "Hello, World!\x0A", 14) = 14
qemu-system-x86_64: terminating on signal 2
```

This will run a dynamically linked `helloworld` application.
You will see a lot of output because, by default, the `run_app.sh` script uses a prebuilt version of `app-elfloader`, which has `strace`-like output enabled, for debugging purposes.
You can see a list of all existing available application that you can run in binary compatibility mode by running `./run_app.sh` with no arguments.
**Note: If you are running an application that requires networking support, you will need to run it with `sudo`.***

> If you want more control over the application, you can use the `run.sh` script.

Reviewer: This is too much. I would move this to section 1.5 (i.e. between current sections 1 and 2, with a name such as "Behind the Scenes"). The "storyline" would be like:

* `00. Setup`: Quick setting up of binary compatibility. Have everything ready, run helloworld.

* `01. Run Binary Applications`: Run all the applications for binary compatibility. Use `run_app.sh`.

* `02: Flavors of ELF Loader`: Detail existing pre-built version of `elfloader` and updating them via the `defaults` file. Continue using `run_app.sh`.

* `03. Behind the Scenes`: Introduce the `run.sh` script, run commands using `run.sh`. Show filesystem information in the `dynamic-apps` / `static-apps` repository. The contents detailing `run.sh` from this setup section together with contents from current 03 section will end up here.

* `04. Run New Applications`: Use `extract.sh` script or a manual approach to create a filesystem for a simple CLI application already installed (or that needs to be installed) in the system (`tree` or `zip` or `unzip` or anything dealing with the filesystem - it's easier; for processes, you require `procfs` support and that's more difficult).

* `05. Build and Run Custom Applications`: Provide a sample source code or link to a repository where they would build and run both static and dynamic apps. I would give two subtasks. A simple one, from some toy repository (something like [my `snippets` repository](https://github.com/razvand/snippets)), with an existing `Makefile` with rules to build a static and dynamic application. And then to a practical one (I would go for a Go or Rust repository) that they clone, build, run locally and the run via the `elfloader`.
  The thing is that build the ELF Loader is not that relevant. But running different applications, building them, that's more practical.

* `06: Build app-elfloader`: First show a full command to run `helloworld` and `nginx`. Show the usefulness of the `run_app.sh` and the `run.sh` script., that hides the complexity of using `qemu-system-x86_64`. Then proceed to build `app-elfloader`.

* `07: Debugging Binary Compatibility`: This is a bonus item and can be left out. I would ask them to run a process/memory/user-related CLI executable (`free`, `id` etc.). And see that it requires a certain filesystem (probably files in `/etc/` or `/proc/`) and ask them to create the proper filesystem for it to run.

> The `elfloader` application will need to be passed a filesystem that contains all the dependencies (i.e. external libraries) the already compiled application needs, along with the [system loader](https://www.man7.org/linux/man-pages/man8/ld.so.8.html), and will expect as an argument the name of the `ELF` that will be executed.

Reviewer: This is a convoluted phrase. Break it in smaller sentences, reword where necessary (i.e. `need` / `needs` appears twice, `will` appears multiple times).

The `dynamic-apps` repository already has those prepared, as you can see below:

```console

> $ cd path/to/dynamic-apps-repo

Reviewer: ```suggestion
cd ../dynamic-apps
```

> $ ldd lang/c++/helloworld

Reviewer: Give up `$` signs, detach command from its output.

    linux-vdso.so.1 (0x00007ffd60841000)
    libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fe9f27d3000)
    libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fe9f25e1000)
    libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fe9f2492000)
    /lib64/ld-linux-x86-64.so.2 (0x00007fe9f2c1e000)
    libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fe9f2477000)

$ tree lang/c++/
lang/c++/
|-- helloworld
|-- helloworld.cpp
|-- lib
|   `-- x86_64
|       |-- libc.so.6
|       |-- libgcc_s.so.1
|       `-- libm.so.6
|-- lib64
|   `-- ld-linux-x86-64.so.2
|-- Makefile
`-- usr
    `-- lib
        `-- x86_64-linux-gnu
            `-- libstdc++.so.6

6 directories, 8 files
```

You can see the dependencies of an application by running [`ldd`](https://www.man7.org/linux/man-pages/man1/ldd.1.html).
As you can see, all the dependencies needed to run `apphelloworld` are located in the root `c++/` directory.

> You can ignore the `linux-vdso.so.1` for now.

Reviewer: Mingle these explanations with the commands. Do not show explanations at the end, show them inline.

> In order to run the c++ `helloworld` application using the Unikraft `elfloader`, we will pass the `c++/` directory we saw above to the application as a filesystem, along with the `/helloworld` argument, which will represent the path to the `ELF` file.

Reviewer: ```suggestion
In order to run the C++ `helloworld` application using the Unikraft `elfloader`, we will pass the `c++/` directory we saw above to the application as a filesystem, along with the `/helloworld` argument, which will represent the path to the `ELF` file.
```

Reviewer: Break this phrase.

> We can do this by running `qemu-system-x86_64`, like we did in the previous sessions, but we can also use the `run.sh` helper script from the `run-app-elfloader` repository:

Reviewer: Use two sentences.

```console

> $ ./run.sh -r ../dynamic-apps/lang/c++/ "/helloworld"

Reviewer: Remove quotes, it appears weird.
